### ISSUE TITLE
Incorrect expiration time rendering with PHP 8.1 bug

### ISSUE THREAD

**Issue by komputerwiz** [SymfonyCasts/reset-password-bundle#219]
Expiration time is once again not being rendered correctly. A new Symfony 6 project with `make:reset-password` displays the following:

> If an account matching your email exists, then an email was just sent that contains a link that you can use to reset your password. This link will expire in 11 months.
>
> If you don't receive an email please check your spam folder or try again.

The reset token lifetime is 3600 by default, and the behavior is the same if set explicitly.

This issue appears to have been reported in the past (#146), but I'm opening a new ticket since the code has been updated to use translations.

The problem appears to arise from a bad interval that results from time zone confusion (this is probably a bug in PHP). I have distilled the problem to the following snippet, which is what [ResetPasswordHelper:71](https://github.com/SymfonyCasts/reset-password-bundle/blob/6e926da7fb2031ef38a385a994ea3a52b436dc8c/src/ResetPasswordHelper.php#L71) and [ResetPasswordToken:141](https://github.com/SymfonyCasts/reset-password-bundle/blob/6e926da7fb2031ef38a385a994ea3a52b436dc8c/src/Model/ResetPasswordToken.php#L141) are doing under the hood:

```php
// creates new DateTime for the current timezone (UTC-5 in my case)
$expiresAt = new \DateTimeImmutable('+3600 seconds'); 
echo 'expires: ' . $expiresAt->format(\DateTimeInterface::ISO8601) . "\n";

$generatedAt = ($expiresAt->getTimestamp() - 3600);

// creates new DateTime in UTC
$createdAtTime = \DateTimeImmutable::createFromFormat('U', (string) $generatedAt);
echo 'created: ' . $createdAtTime->format(\DateTimeInterface::ISO8601) . "\n";

$interval = $expiresAt->diff($createdAtTime);
echo 'computed interval: ' . $interval->format('%R[%yy %mM %dd %hh %im %ss]') . "\n";
```

This snippet outputs, for example

```
expires: 2022-03-21T19:07:42-0500
created: 2022-03-21T23:07:42+0000
computed interval: -[-1y 11M 31d 0h 0m 0s]
```

Setting the `$expiresAt` timezone to UTC produces the correct result:

```
expires: 2022-03-22T00:11:56+0000
created: 2022-03-21T23:11:56+0000
computed interval: -[0y 0M 0d 1h 0m 0s]
```

Hence, one possible solution would be to enforce a UTC timezone for all dates. There was a similar workaround when the expiration interval was rendered in templates using the `date` Twig filter. (e.g., `expiresInterval|date('g', 'UTC')`). However, an analogous template-driven workaround is no longer accessible since the label generation happens within the ResetPasswordToken class itself when generating the translation message key and data.

My advice is to change the expiration DateTime generation at [ResetPasswordHelper:71](https://github.com/SymfonyCasts/reset-password-bundle/blob/6e926da7fb2031ef38a385a994ea3a52b436dc8c/src/ResetPasswordHelper.php#L71) to the following:

```php
$expiresAt = new \DateTimeImmutable(sprintf('+%d seconds', $this->resetRequestLifetime), new \DateTimeZone('UTC'));
```

I can submit a pull request if you would like.

**Comment by jrushlow** [comment]
Howdy @komputerwiz! Going based on your code snippet above (Modified the `DateTimeImmutable::_construct` to use -5 timezone): 
```php
// creates new DateTime for the current timezone (UTC-5 in my case)
$expiresAt = new \DateTimeImmutable('+3600 seconds', new \DateTimeZone('-5')); 
echo 'expires: ' . $expiresAt->format(\DateTimeInterface::ISO8601) . "\n";

$generatedAt = ($expiresAt->getTimestamp() - 3600);

// creates new DateTime in UTC
$createdAtTime = \DateTimeImmutable::createFromFormat('U', (string) $generatedAt);
echo 'created: ' . $createdAtTime->format(\DateTimeInterface::ISO8601) . "\n";

$interval = $expiresAt->diff($createdAtTime);
echo 'computed interval: ' . $interval->format('%R[%yy %mM %dd %hh %im %ss]') . "\n";
```

Produces: (https://3v4l.org/AZpkv)
![image](https://user-images.githubusercontent.com/40327885/159501434-1eef3d67-ac20-4f71-aefd-a13df61da4f5.png)

This appears to be a bug related to PHP `8.1.x` but we'll need to do a bit more research to confirm. I'm labeling this as "unconfirmed" as it is indeed a problem - but not directly with our code base. I think the best approach would be to attempt to get this fixed in PHP 8.1 and then determine if mitigation is needed in `reset-password-bundle` to handle the edge cases between PHP `8.1.0` -> `8.1.4`.

**Comment by jrushlow** [comment]
For reference, this has been reported upstream. https://bugs.php.net/bug.php?id=81263

**Comment by komputerwiz** [comment]
Thanks for the quick response!

Yeah, I am using PHP 8.1. In hindsight, I should have reported my platform specs. For the record:

- **Operating System:** Ubuntu 20.04, Linux kernel 5.3.0 x86_64; not virtualized
- **PHP version:** PHP 8.1.3 (CLI) from [ppa:ondrej/php](https://launchpad.net/~ondrej/+archive/ubuntu/php)

Good find on the upstream bug report! I had a strong feeling that this was a PHP issue and not with your code since time zones really only matter for display and not math. Hopefully a fix comes out soon, in which case mitigation probably won't be necessary.

That 3v4l.org site is really neat, by the way. I'll have to save that for later.

Thanks again for taking the time to look at this. Let me know if I can be of any help.

**Comment by komputerwiz** [comment]
To follow up, I think I found the root of the problem in the PHP source code and added a comment to the PHP bug report linked above.

**Comment by acirulis** [comment]
Hi, incorrect expiration interval in email template appeared once my time zone (Europe/Riga) switched to DST on 2022-03-27. I have been using same PHP 8.1 version last week and this week, but problem started only after switching to DST.

**Comment by Tofandel** [comment]
Wow I was wondering why it was saying the token expires in `0 minutes` and scratching my head for why the config option wasn't applying guess that's why..

**Comment by Tofandel** [comment]
Can we maybe get a patch here depending on the php version until this is fixed?

As php version don't get updated nearly as fast as deps

**Comment by WebDevM4DC** [comment]
Now running on PHP 8.1.7 with the same issue.

**Comment by jrushlow** [comment]
There is a fix upstream in PHP https://github.com/php/php-src/issues/8730, we're waiting for `timelib` to get merged into `php` and this should be resolved.

**Comment by wvandenhaak** [comment]
A possible solution, regardless of the mentioned PHP bug fix, is to change the date diff calculations from Unix timestamp to a DateTime(Immutable).

I have applied the following patch to my code, which proved to be a solution. In both PHP versions (7.4.30 and 8.1.7) the expiration time is shown as "30 minutes" (I had the lifetime set  to 1800 seconds).
Without the patch, on PHP 8.1.7, 11 months was shown as expiration time.

I can submit a pull request if you would like.

```diff
diff --git a/src/Model/ResetPasswordToken.php b/src/Model/ResetPasswordToken.php
index 46f6767..8d60d0c 100644
--- a/src/Model/ResetPasswordToken.php
+++ b/src/Model/ResetPasswordToken.php
@@ -26,7 +26,7 @@ final class ResetPasswordToken
     private $expiresAt;
 
     /**
-     * @var int|null timestamp when the token was created
+     * @var \DateTimeImmutable|null timestamp when the token was created
      */
     private $generatedAt;
 
@@ -35,7 +35,7 @@ final class ResetPasswordToken
      */
     private $transInterval = 0;
 
-    public function __construct(string $token, \DateTimeInterface $expiresAt, int $generatedAt = null)
+    public function __construct(string $token, \DateTimeInterface $expiresAt, \DateTimeImmutable $generatedAt = null)
     {
         $this->token = $token;
         $this->expiresAt = $expiresAt;
@@ -138,9 +138,7 @@ final class ResetPasswordToken
             throw new \LogicException(sprintf('%s initialized without setting the $generatedAt timestamp.', self::class));
         }
 
-        $createdAtTime = \DateTimeImmutable::createFromFormat('U', (string) $this->generatedAt);
-
-        return $this->expiresAt->diff($createdAtTime);
+        return $this->expiresAt->diff($this->generatedAt);
     }
 
     private function triggerDeprecation(): void
diff --git a/src/ResetPasswordHelper.php b/src/ResetPasswordHelper.php
index a6b2b6a..bda69e5 100644
--- a/src/ResetPasswordHelper.php
+++ b/src/ResetPasswordHelper.php
@@ -68,9 +68,8 @@ class ResetPasswordHelper implements ResetPasswordHelperInterface
             throw new TooManyPasswordRequestsException($availableAt);
         }
 
-        $expiresAt = new \DateTimeImmutable(sprintf('+%d seconds', $this->resetRequestLifetime));
-
-        $generatedAt = ($expiresAt->getTimestamp() - $this->resetRequestLifetime);
+        $generatedAt = new \DateTimeImmutable();
+        $expiresAt = $generatedAt->modify(sprintf('+%d seconds', $this->resetRequestLifetime));
 
         $tokenComponents = $this->tokenGenerator->createToken($expiresAt, $this->repository->getUserIdentifier($user));

```

**Comment by micayael** [comment]
Same here with lifetime: 3600

PHP 8.1.8 (cli) (built: Jul 11 2022 08:29:57) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.8, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.8, Copyright (c), by Zend Technologies


[ResetPasswordToken.php](https://localhost:8001/_profiler/open?file=vendor/symfonycasts/reset-password-bundle/src/Model/ResetPasswordToken.php&line=142#line142) on line 142:
DateInterval {#1364 ▼
  interval: - -1y 11m 31d 01:00:00.858284
  +"y": -1
  +"m": 11
  +"d": 31
  +"h": 1
  +"i": 0
  +"s": 0
  +"f": 0.858284
  +"weekday": 0
  +"weekday_behavior": 0
  +"first_last_day_of": 0
  +"invert": 1
  +"days": 0
  +"special_type": 0
  +"special_amount": 0
  +"have_weekday_relative": 0
  +"have_special_relative": 0
}

**Comment by ganti** [comment]
thanks for preparing the fix #235 

PHP 8.1.9 was released just a few days ago. They have more or less a [monthly cycle](https://www.php.net/ChangeLog-8.php#8.1.9). therefore it will be roughly begin of September until PHP 8.1.10 will be released.

**Comment by ToshY** [comment]
Just tested it with `8.1.10`, works again 🥳.

**Comment by Machou** [comment]
The fix by @wvandenhaak Thanks !

Work with PHP 8.1.5

**Comment by Raghav9888** [comment]
composer require symfonycasts/reset-password-bundle
Using version ^1.15 for symfonycasts/reset-password-bundle
./composer.json has been updated
Running composer update symfonycasts/reset-password-bundle
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - php is present at version 8.1.0 and cannot be modified by Composer
    - symfonycasts/reset-password-bundle v1.15.0 conflicts with php >=8.1 <8.1.10.
    - Root composer.json requires symfonycasts/reset-password-bundle ^1.15 -> satisfiable by symfonycasts/reset-password-bundle[v1.15.0].

You can also try re-running composer require with an explicit version constraint, e.g. "composer require symfonycasts/reset-password-bundle:*" to figure out if any version is installable, or "composer require symfonycasts/reset-password-bundle:^2.1" if you know which you need.

Installation failed, reverting ./composer.json and ./composer.lock to their original content.


how to remove this error

**Comment by Tofandel** [comment]
@Raghav9888 Update your system's php version or if you don't care about the bug, install an older version of the bundle

**Comment by Raghav9888** [comment]
[https://prnt.sc/JtwMsdjxwtsb](url)
Hello Sir 
I am student please help me

**Comment by Raghav9888** [comment]
/**
 * @psalm-immutable 
 */
  interface Type{  

 *   const EntityType = ;
 * 
    /**  
   * Returns a rendered output of the Type as it would be used in a DocBlock. 
   * */  
  public function __toString(): string;

**Comment by Tofandel** [comment]
@Raghav9888 you are spamming the thread of a closed issue with something that has nothing to do with it.

You seem to be editing files in the /vendor directory. Which you should never do

Delete the vendor folder and rerun composer install

If the issue is still there Google the error message. Open an issue in the relevant github repository if nothing turns up

**Comment by Raghav9888** [comment]
Can you tell me when we Use inheritance

abstract class AbstractEntity ( #[ORM\MappedSuperclass])
class LoginUser
(
#[ORM\Entity(repositoryClass: LoginUserRepository::class)]
#[ORM\InheritanceType("SINGLE_TABLE")]
#[ORM\DiscriminatorColumn(name:"discr", type:"string")]
#[ORM\DiscriminatorMap(['loginUser'=>LoginUser::class,"loginUserDetails"=>LoginUserDetails::class])]
)
3)class LoginUserDetails extends LoginUser
how to create form ?

**Comment by Raghav9888** [comment]
Any One help me please tell me. How to fix 
**Failed to send an update.**  this error
and this error **mercure?topic=chat**

**Comment by jrushlow** [comment]
@Raghav9888 - this is the repository for `ResetPasswordBundle`, you'll need to seek help for `mercure` in the Mercure repo at https://github.com/dunglas/mercure
